src/System.IO.FileSystem/tests/FileStream/ReadAsync.cs

```diff
@@ -0,0 +1,329 @@
+
+                    Array.Clear(buffer, 0, buffer.Length);
+
+                    // read should now complete synchronously since it is serviced by the read buffer filled in the first request
+                    Assert.Equal(TestBuffer.Length,
+                        FSAssert.CompletesSynchronously(fs.ReadAsync(buffer, 0, buffer.Length), checkCompletedSynchronously: false));
+                    Assert.Equal(TestBuffer, buffer);
+                }
+            }
+        }
+
+        [Fact]
+        public async Task ReadAsyncExistingFile()
+        {
+            string fileName = GetTestFilePath();
+            using (FileStream fs = new FileStream(fileName, FileMode.Create))
+            {
+                fs.Write(TestBuffer, 0, TestBuffer.Length);
+            }
+
+            using (FileStream fs = new FileStream(fileName, FileMode.Open))
+            {
+                byte[] buffer = new byte[TestBuffer.Length];
+                Assert.Equal(TestBuffer.Length, await fs.ReadAsync(buffer, 0, buffer.Length));
+                Assert.Equal(TestBuffer, buffer);
+
+                // ReadAsync with too large buffer at front of buffer
+                fs.Position = 0;
+                buffer = new byte[TestBuffer.Length * 2];
+                Assert.Equal(TestBuffer.Length, await fs.ReadAsync(buffer, 0, buffer.Length));
+                Assert.Equal(TestBuffer, buffer.Take(TestBuffer.Length));
+                // Remainder of buffer should be untouched.
+                Assert.Equal(new byte[buffer.Length - TestBuffer.Length], buffer.Skip(TestBuffer.Length));
+
+                // ReadAsync with too large buffer in middle of buffer
+                fs.Position = 0;
+                buffer = new byte[TestBuffer.Length * 2];
+                Assert.Equal(TestBuffer.Length, await fs.ReadAsync(buffer, 2, buffer.Length - 2));
+                Assert.Equal(TestBuffer, buffer.Skip(2).Take(TestBuffer.Length));
+                // Remainder of buffer should be untouched.
+                Assert.Equal(new byte[2], buffer.Take(2));
+                Assert.Equal(new byte[buffer.Length - TestBuffer.Length - 2], buffer.Skip(2 + TestBuffer.Length));
+            }
+        }
+
+        [Fact]
+        public async Task ReadAsyncCancelledFile()
+        {
+            string fileName = GetTestFilePath();
+            using (FileStream fs = new FileStream(fileName, FileMode.Create))
+            {
+                while(fs.Length < 128 * 1024)
+                    fs.Write(TestBuffer, 0, TestBuffer.Length);
+            }
+
+            using (FileStream fs = new FileStream(fileName, FileMode.Open))
+            {
+                byte[] buffer = new byte[fs.Length];
+                CancellationTokenSource cts = new CancellationTokenSource();
+                Task<int> readTask = fs.ReadAsync(buffer, 0, buffer.Length, cts.Token);
+                cts.Cancel();
+                try
+                {
+                    await readTask;
+                }
```

Comment:
You probably want an Assert.True(false) or something similar here after the await readTask to make sure that it actually throws.